Identify UTxOs suitable for use as collateral

sevanspowell opened the pull request:
# Issue Number

ADP-1053


# Overview

- Added a function `asCollateral` to identify UTxOs suitable as collateral inputs.
- Clarified the meaning of a "VK input".
- Thoroughly tested the collateral functions.
- Added a property test to test the behaviour of `TokenBundle.toCoin`.

# Comments
[
I've clarified with the ledger team](https://input-output-rnd.slack.com/archives/CCRB7BU8Y/p1628060541075300) that a UTxO is to be considered suitable for collateral iff the payment credential associated with the output address of the UTxO is of type "key hash".

See the "Binary Address Format" heading of this spec: https://hydra.iohk.io/build/6752483/download/1/ledger-spec.pdf

sevanspowell committed:
Test behaviour of TokenBundle.toCoin

- Add a property test to verify the behaviour of TokenBundle.toCoin.

sevanspowell committed:
Identify UTxO entries suitable for use as collateral

- Add function "asCollateral", that is used to determine if a given UTxO is
  suitable for use as a collateral input. For a UTxO to be considered a suitable
  collateral input, it must:
     - Be a pure ADA UTxO (no tokens)
     - Require a verification key witness to be spent
     - Have an output address that is not any of:
       - a native script address
       - a plutus script address
- Thoroughly test collateral suitability functions.

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

sevanspowell commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/CollateralSpec.hs:
```diff
@@ -0,0 +1,810 @@
+    B.runGet getHeader delegationAddrGolden
+        `shouldBe` BaseAddress CredentialKeyHash CredentialKeyHash
+
+-- TODO generate more unit tests for each type of address.
```
These were the only types of addresses I could generate with cardano-addresses. I would have liked to generate one of each of the 11 different types of addresses. Does anyone know how I might do that?

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Collateral.hs:
```diff
@@ -0,0 +1,105 @@
+    -> Maybe Coin
+    -- ^ The total ADA value of that UTxO if it is suitable for collateral,
+    -- otherwise Nothing.
+asCollateral (_txIn, txOut) = do
```
Since the `TxIn` argument is ignored, what do you think about simplifying this function's type, to:
```hs
asCollateral :: TxOut -> Maybe Coin
```
This would remove the need to generate arbitrary `TxIn` values.

Then where this function is used, we can call `Map.mapMaybe` with `asCollateral` as the first argument:

```hs
utxoCollateral :: UTxO -> Map TxIn Coin
utxoCollateral (UTxO u) = Map.mapMaybe asCollateral u
```

What do you think?

I've added a commit here which demonstrates the suggestion: cdf4d12b2dba144b6f9fde544f2b30aaa835e913

sevanspowell replied:
I've cherry-picked that commit, sounds good :+1:

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Collateral.hs:
```diff
@@ -0,0 +1,105 @@
+    deriving (Eq, Show)
+
+-- | Analyze an address to determine if it's funds are suitable for use as a
+-- collateral input.
```
```suggestion
-- | Analyze an address to determine if its funds are suitable for use as a
-- collateral input.
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Collateral.hs:
```diff
@@ -0,0 +1,105 @@
+    -- ^ The address is some form of stake address
+    | IsAMalformedOrUnknownAddr
+    -- ^ The address could not be parsed
+    deriving (Eq, Show)
```
Just a suggestion (shorter names):
```suggestion
data AddrNotSuitableForCollateral
    = IsScriptAddr
    -- ^ The address is some form of script address
    | IsStakeAddr
    -- ^ The address is some form of stake address
    | IsMalformedOrUnknownAddr
    -- ^ The address could not be parsed
    deriving (Eq, Show)
```

sevanspowell replied:
Done :smiley:

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Collateral.hs:
```diff
@@ -0,0 +1,105 @@
+--
+-- This function returns an Either instead of a Maybe because I think it's
+-- important these functions return why they failed. It is extremely useful for
+-- debugging programs.
```
How about:

```suggestion
-- This function returns an Either instead of a Maybe so that we can test
-- that it has failed for the right reason.
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Collateral.hs:
```diff
@@ -0,0 +1,105 @@
+                    Left IsAScriptAddr
+
+                -- Otherwise, this is an address that corresponds with a
+                -- suitable collateral input
```
```suggestion
                -- Otherwise, this is an address that is suitable to be
                -- a collateral input
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,810 @@
+-- | Bootstrap address  | keyhash            | -                |                 1000 |
+-- | Stake address      | -                  | keyhash          |                 1110 |
+-- |                    | -                  | scripthash       |                 1111 |
+-- | Future formats     | ?                  | ?                |            1001-1101 |
```
Nice table!

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,810 @@
+-- test.
+--
+-- First we write our generator, which customizes it's frequency so that every
+-- branch of the AddressType type is evenly covered.
```
```suggestion
-- First we write our generator, which customizes its frequency so that every
-- branch of the AddressType type is evenly covered.
```

jonathanknowles committed:
Extract out header nibble encoding/decoding into dedicated functions.

This commit separates out the following responsibilities:

- defining the bidirectional mapping between an address type and its
  corresponding header nibble
- reading or modifying the address type of an address

jonathanknowles committed:
Simplify type of function `asCollateral`.

This commit simplifies the type of `asCollateral` to just take a `TxOut`.

The value of the `TxIn` parameter was always ignored, so it can be
safely removed from the input of the function.

This enables us to make some simplifications to the test suite.

It won't affect the usability of the `asCollateral` function, since we can
now call it like this:

```hs
utxoCollateral :: UTxO -> Map TxIn Coin
utxoCollateral (UTxO u) = Map.mapMaybe asCollateral u
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,810 @@
+        0b10000000 -> pure BootstrapAddress
+        0b11100000 -> pure $ StakeAddress CredentialKeyHash
+        0b11110000 -> pure $ StakeAddress CredentialScriptHash
+        _          -> fail "Unknown address type."
```
Perhaps we can separate out the following responsibilities:
    
- defining the bidirectional mapping between an address type and its corresponding header nibble
  (this can be expressed in terms of **`Word4`**)
- reading or modifying the address type of an address

I've made a commit to demonstrate the idea here: 8f338d153140bd8c42d43df78bfbbee69bf0967d

What do you think?

sevanspowell replied:
That's great!

jonathanknowles reviewed (commented):
Hi @sevanspowell 

I've not finished reviewing everything, but this looks good so far. 👍🏻 

I've made a few suggestions (see comments).

For two of the suggestions, I coded them up in a separate PR (https://github.com/input-output-hk/cardano-wallet/pull/2809). If you like these suggestions, feel free to `git cherry-pick` them!

I'll come back to do more review a bit later on!

sevanspowell committed:
Shorten names of AddrNotSuitableForCollateral enums

sevanspowell committed:
Re-arrange AddressType and related function

sevanspowell committed:
Improve the quality of property tests using "disjoin"

sevanspowell committed:
Rework asCollateral implementation

- Separate out the concepts of retrieving an address type and determining if
  that address type is suitable for use as collateral.
- Rework the "asCollateral" function in terms of these separated concepts.
- "classifyCollateralAddress" is no longer strictly needed, but is left in as a
  useful test tool, and is already heavily tested.
- Add tests for new separted concepts, with equivalences setup with respect to
  existing tests.
- Re-arrange commentary a bit so it makes more sense.

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Collateral.hs:
```diff
@@ -0,0 +1,248 @@
+    B.putWord8 $
+    fromIntegral @Word4 @Word8 (addressTypeToHeaderNibble t) `Bits.shiftL` 4
+
+-- By inspecting the bit pattern of an Address, we can determine it's address
```
```suggestion
-- By inspecting the bit pattern of an Address, we can determine its address
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Collateral.hs:
```diff
@@ -0,0 +1,248 @@
+         False
+
+    BootstrapAddress ->
+         True
```
This is one of those (rare) situations where I'd advocate for using alignment in the pattern match:
- Pre-condition: the resulting code does not overflow the margin (80 columns).
- Justification: the pattern match really is tabular in this case. It's arguably easier to read and understand if it's laid out in a tabular fashion.

I'd also recommend avoiding wildcards here:
- Justification: if someone needs to add a constructor to `Credential` in future, then the compiler will complain about every single one of the missing matches, rather than just a subset, and we can avoid accidental matches with new constructors that might possibly give the wrong answer!

```suggestion
    BaseAddress CredentialKeyHash CredentialKeyHash       -> True
    BaseAddress CredentialKeyHash CredentialScriptHash    -> True
    BaseAddress CredentialScriptHash CredentialKeyHash    -> False
    BaseAddress CredentialScriptHash CredentialScriptHash -> False
    PointerAddress CredentialKeyHash                      -> True
    PointerAddress CredentialScriptHash                   -> False
    EnterpriseAddress CredentialKeyHash                   -> True
    EnterpriseAddress CredentialScriptHash                -> False
    StakeAddress CredentialKeyHash                        -> False
    StakeAddress CredentialScriptHash                     -> False
    BootstrapAddress                                      -> True
```
```

jonathanknowles committed:
Add function `addressSuitableForCollateral`.

This is a simple composition of `addressType` and
`addressTypeSuitableForCollateral`.

jonathanknowles committed:
Use `addressSuitableForCollateral` to simplify `asCollateral`.

jonathanknowles committed:
Use `addressSuitableForCollateral` to simplify `CollateralSpec`.

jonathanknowles committed:
Remove `classifyCollateralAddress` function.

This function is now redundant, given the existence of
`addressSuitableForCollateral`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Collateral.hs:
```diff
@@ -0,0 +1,248 @@
+classifyCollateralAddress
+    :: Address
+    -> Either AddrNotSuitableForCollateral Address
+classifyCollateralAddress addr@(Address addrBytes) =
```
I think there is a good simplification available to us here:

It seems to me that `classifyCollateralAddress` and `addressTypeSuitableForCollateral` are duplicating some of the same logic. (Though please correct me if you think I'm wrong about this, or have missed something!)

Suppose we define the following function:
```hs
addressSuitableForCollateral :: Address -> Bool
addressSuitableForCollateral =
    maybe False addressTypeSuitableForCollateral . addressType
```
Then we can use this function wherever we would have used `classifyCollateralAddress` (more or less), provided we are happy with the classification that `addressType` is already performing (and it is tested).

This also allows us to remove the `classifyCollateralAddress` function and associated `AddrNotSuitableForCollateral` type, with no change in the behaviour, and no loss of functionality, with only some minor tweaks to the test suite.

I've coded up this suggestion in a PR here: https://github.com/input-output-hk/cardano-wallet/pull/2814

What do you think?

sevanspowell replied:
Yeah, I think that's good. The only reason I kept it was so that debugging is a bit easier (it tells you why, rather than just "no"). But it's not a super important point, I like your PR :+1:

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+genCredential =
+    oneof [ pure CredentialKeyHash
+          , pure CredentialScriptHash
+          ]
```
Use 4-char indents rather than alignment here: (https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#avoid-variable-length-indentation)

```suggestion
genCredential :: Gen Credential
genCredential = oneof
    [ pure CredentialKeyHash
    , pure CredentialScriptHash
    ]
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+        , ("StakeAddress CredentialScriptHash"                     , 5)
+        , ("BootstrapAddress"                                      , 5)
+        ] $
+        tabulate "Address types" [show addrType] $ True === True
```
👍🏻

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+            addrType = B.runGet getAddressType (BL.fromStrict addrBytes)
+        coverTable "Address types"
+            [ ("StakeAddress CredentialKeyHash"                        , 30)
+            , ("StakeAddress CredentialScriptHash"                     , 30)
```
```suggestion
            [ ("StakeAddress CredentialKeyHash"   , 30)
            , ("StakeAddress CredentialScriptHash", 30)
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+            disjoin
+              [ addrType === StakeAddress CredentialKeyHash
+              , addrType === StakeAddress CredentialScriptHash
+              ]
```
```suggestion
            disjoin
                [ addrType === StakeAddress CredentialKeyHash
                , addrType === StakeAddress CredentialScriptHash
                ]
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+              , addrType === BaseAddress CredentialKeyHash CredentialScriptHash
+              , addrType === PointerAddress CredentialKeyHash
+              , addrType === EnterpriseAddress CredentialKeyHash
+              ]
```
```suggestion
                [ addrType === BaseAddress CredentialKeyHash CredentialKeyHash
                , addrType === BaseAddress CredentialKeyHash CredentialScriptHash
                , addrType === PointerAddress CredentialKeyHash
                , addrType === EnterpriseAddress CredentialKeyHash
                ]
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+                === PointerAddress CredentialScriptHash
+              , addrType
+                === EnterpriseAddress CredentialScriptHash
+              ]
```
```suggestion
                [ addrType
                  === BaseAddress CredentialScriptHash CredentialKeyHash
                , addrType
                  === BaseAddress CredentialScriptHash CredentialScriptHash
                , addrType
                  === PointerAddress CredentialScriptHash
                , addrType
                  === EnterpriseAddress CredentialScriptHash
                ]
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+              , (1, asAddress <$> genByronAddr)
+              , (2, asStakeAddress <$> genStakeAddr)
+              , (3, Address <$> arbitrary)
+              ]
```
Avoid variable-spaced alignment here: https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#avoid-variable-length-indentation
```suggestion
genAnyAddress = frequency
    [ (10, asAddress <$> genShelleyAddr)
    , (1, asAddress <$> genByronAddr)
    , (2, asStakeAddress <$> genStakeAddr)
    , (3, Address <$> arbitrary)
    ]
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+                "could simplify address"  $
+            case simplifyAddress addr of
+                Nothing ->
+                    True === True
```
Here you can just use the `property` function:
```suggestion
                    property True
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+          :: Maybe (L.Addr CC.StandardCrypto))
+  ||
+  isJust (L.deserialiseRewardAcnt addrBytes
+          :: Maybe (L.RewardAcnt CC.StandardCrypto))
```
Use 4-char indentations here:
```suggestion
    isJust (L.deserialiseAddr addrBytes
        :: Maybe (L.Addr CC.StandardCrypto))
    ||
    isJust (L.deserialiseRewardAcnt addrBytes
        :: Maybe (L.RewardAcnt CC.StandardCrypto))
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+genShelleyScriptHashAddr =
+    L.Addr <$> arbitrary
+           <*> (L.ScriptHashObj . L.ScriptHash <$> L.genHash)
+           <*> arbitrary
```
https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#avoid-variable-length-indentation
```suggestion
genShelleyScriptHashAddr = L.Addr
    <$> arbitrary
    <*> (L.ScriptHashObj . L.ScriptHash <$> L.genHash)
    <*> arbitrary
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+genShelleyKeyHashAddr =
+    L.Addr <$> arbitrary
+           <*> (L.KeyHashObj <$> arbitrary)
+           <*> arbitrary
```
https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#avoid-variable-length-indentation
```suggestion
genShelleyKeyHashAddr = L.Addr
    <$> arbitrary
    <*> (L.KeyHashObj <$> arbitrary)
    <*> arbitrary
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/TokenBundleSpec.hs:
```diff
@@ -173,6 +177,24 @@ prop_equipartitionQuantitiesWithUpperBound_sum m maxQuantity =
+            "Token bundle has no non-ada assets" $
+        if TokenBundle.isCoin bundle
+        then result === Just (TokenBundle.coin bundle)
+        else result === Nothing
```
👍🏻

jonathanknowles reviewed (commented):
Hi @sevanspowell 

Many thanks again for writing this PR! I've nearly finished reviewing everything, but still have a bit more to read.

I've left some comments and suggestions. Hopefully these make sense, but let me know if not!

In particular, I think there's a simplification possible which can remove some of the functions without loss in functionality. (See [this comment](https://github.com/input-output-hk/cardano-wallet/pull/2807#discussion_r685707721)). Would be interested to hear your thoughts.

sevanspowell committed:
Made a few tweaks around suggested improvements

- Took Jonathan's changed property and made a new property, restoring the old
one, as they are slightly different, Bool =/= Maybe a.
- Cleared up and updated some of the commentary.

sevanspowell committed:
Formatting

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+                "could simplify address"  $
+            case simplifyAddress addr of
+                Nothing ->
+                    True === True
```
Suggestion: use `property` here:
```suggestion
                    property True
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+                    in
+                        originalAddressType === simplifiedAddressType
+
+-- From this function we can generator a QuickCheck shrinker:
```
```suggestion
-- From this function we can generate a QuickCheck shrinker:
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+-- cat recovery-phrase.txt | cardano-address key from-recovery-phrase Byron > root.prv
+-- cat root.prv | cardano-address key child 14H/42H | tee addr.prv | cardano-address key public --with-chain-code | cardano-address address bootstrap --root $(cat root.prv | cardano-address key public --with-chain-code) --network-tag testnet 14H/42H
+byronAddrGolden :: BL.ByteString
+byronAddrGolden = BL.fromStrict . fromJust . decodeBase58 bitcoinAlphabet $ "37btjrVyb4KFsMoVwPRZ5aJko48uBFFUnJ46eV3vC3uBCC65mj5BfbGP6jYDfhojm8MAayHo4RPvWH4x852FcJq8SHazCx31FJM2TfDpV9Azrc8UKD"
```
This might make it easier to read in an editor:

```suggestion
byronAddrGolden = BL.fromStrict . fromJust . decodeBase58 bitcoinAlphabet $ 
    "37btjrVyb4KFsMoVwPRZ5aJko48uBFFUnJ46eV3vC3uBCC65mj5BfbGP6jYDfhojm8MAayHo4RPvWH4x852FcJq8SHazCx31FJM2TfDpV9Azrc8UKD"
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+            case addrType of
+                -- Only unrecognized addresses are classified as malformed
+                -- or unknown (i.e. we otherwise classify any known address
+                -- according to it's type)
```
```suggestion
                -- according to its type)
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+-- cat root.prv | cardano-address key child 1852H/1815H/0H/0/0 > addr.prv
+-- cat addr.prv | cardano-address key public --with-chain-code | cardano-address address payment --network-tag testnet | cardano-address address delegation $(cat stake.prv | cardano-address key public --with-chain-code)
+delegationAddrGolden :: BL.ByteString
+delegationAddrGolden = unsafeBech32Decode "addr_test1qpdylg53ekxh2404mfgw4pt4gfm7dc9dkc74ck0gtrld8uyh9dhl3d8cc52zacu6wapdf2pfnepf5n7pp2vfaz7ge8eqd4nn9s"
```
This might make it easier to read in an editor:
```suggestion
delegationAddrGolden = unsafeBech32Decode 
    "addr_test1qpdylg53ekxh2404mfgw4pt4gfm7dc9dkc74ck0gtrld8uyh9dhl3d8cc52zacu6wapdf2pfnepf5n7pp2vfaz7ge8eqd4nn9s"
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -0,0 +1,774 @@
+        , ("StakeAddress CredentialScriptHash"                     , 5)
+        , ("BootstrapAddress"                                      , 5)
+        ] $
+        tabulate "Address types" [show addrType] $ True === True
```
Suggestion: use `property True` here:
```suggestion
        tabulate "Address types" [show addrType] $ property True
```

jonathanknowles reviewed (approved):
Hi Sam

This looks great to me!

There are a few tiny fixes I've proposed. (See comments.)
If you agree with those, feel free to apply them before merging.

Thanks again for all your work on this!
